scarb/src/core/registry/client/http.rs

```diff
@@ -146,12 +151,77 @@ impl<'c> RegistryClient for HttpRegistryClient<'c> {
-        // TODO(mkaput): Publishing to HTTP registries is not implemented yet.
-        Ok(false)
+        let index_config = self.index_config.load().await?;
+        match index_config.upload {
+            Some(_) => Ok(true),
+            None => Ok(false),
+        }
     }
 
-    async fn publish(&self, _package: Package, _tarball: FileLockGuard) -> Result<()> {
-        todo!("Publishing to HTTP registries is not implemented yet.")
+    async fn publish(&self, package: Package, tarball: FileLockGuard) -> Result<RegistryUpload> {
+        let auth_token =
+            env::var("SCARB_AUTH_TOKEN").map_err(|_| anyhow!("Missing authentication token."))?;
+
+        let path = tarball.path().to_owned();
+        // we need to drop, because windows file locking is very strict
+        drop(tarball);
+
+        ensure!(
+            Path::new(&path).exists(),
+            "cannot upload package - file does not exist at path: {}",
+            &path
+        );
+
+        let file = TokioFile::open(&path).await?;
+        let metadata = file.metadata().await?;
+        ensure!(
+            metadata.len() < 5 * 1024 * 1024,
+            "package size ({} B) is too large (max 5MB allowed)",
+            &metadata.len()
+        );
+
+        let index_config = self.index_config.load().await?;
+
+        let file_part = Part::stream(Body::from(file))
+            .file_name(format!("{}_{}", &package.id.name, &package.id.version));
+        let form = Form::new().part("file", file_part);
+
+        let response = self
+            .config
+            .online_http()?
+            .post(
+                index_config
+                    .upload
+                    .clone()
+                    .ok_or_else(|| anyhow!("Upload URL is missing"))?,
+            )
+            .header(AUTHORIZATION, auth_token)
+            .multipart(form)
+            .timeout(Duration::from_secs(60))
+            .send()
+            .await?;
+
+        match response.status() {
+            StatusCode::UNAUTHORIZED => Err(RegistryUpload::Unauthorized)
+                .map_err(|_| anyhow!("Invalid authentication token.")),
+            StatusCode::FORBIDDEN => Err(RegistryUpload::CannotPublish)
+                .map_err(|_| anyhow!("Missing upload permissions or not the package owner.")),
+            StatusCode::BAD_REQUEST => Err(RegistryUpload::VersionExists)
+                .map_err(|_| anyhow!("Package {} already exists.", &package.id)),
+            StatusCode::UNPROCESSABLE_ENTITY => {
+                Err(RegistryUpload::Corrupted).map_err(|_| anyhow!("File corrupted during upload."))
+            }
+            StatusCode::OK => Ok(RegistryUpload::Success),
+            _ => Err(RegistryUpload::Failed).map_err(|_| {
+                anyhow!(
+                    "Upload failed with an unexpected error (trace-id: {:?})",
+                    response
+                        .headers()
+                        .get("x-cloud-trace-context")
```

Comment:
Not sure if it is customary to include in user message 🤔

Reply:
I think we should skip the `(trace-id: ...)` part altogether if data is missing. Otherwise, this makes an impression that something is missing by error.

Reply:
Hope I understood you correctly, here's the change, please take a look: https://github.com/software-mansion/scarb/pull/1570/commits/b23ad3dbf0e027e474bf70343fe558b67499692d